pootle/apps/pootle_store/models.py

```diff
@@ -1697,6 +1697,72 @@ def record_submissions(self, unit, old_target, old_state, current_time, user):
             # `bulk_create()` them in a single go
             sub.save()
 
+    def get_insert_points(self, opcodes, new_unitid_list, old_unitid_list,
+                          old_unitids):
+        """Returns a list of insert points with update index info.
+
+        :param opcodes: result of new and old unitid lists matching.
+        :param new_unitid_list: new list of UIDs.
+        :param old_unitid_list: old list of UIDs.
+        :param old_unitids: dictionary {'dbid': dbid, 'index': index}
+        :return: a list of tuples
+            ``(insert_at, uids_to_add, next_index, update_index_delta)`` where
+            ``insert_at`` is a place for inserting
+            ``uids_to_add`` is units to be inserted
+            ``update_index_delta`` is an offset for index updating
+            ``next_index`` is a place starting from that ``update_index_delta``
+            should be applied.
+        """
+
+        inserts = []
+        for (tag, i1, i2, j1, j2) in opcodes:
+            if tag == 'delete':
+                continue
+            elif tag == 'insert':
+                update_index_delta = 0
+                previous_index = 0
+                if i1 > 0:
+                    previous = old_unitids[old_unitid_list[i1 - 1]]
+                    previous_index = previous['index']
+                next_index = previous_index + 1
+                if i1 < len(old_unitid_list):
+                    next = old_unitids[old_unitid_list[i1]]
+                    next_index = next['index']
+                    delta = j2 - j1
+                    update_index_delta = delta - next_index + previous_index + 1
+
+                inserts.append((previous_index,
+                                new_unitid_list[j1:j2],
+                                next_index,
+                                update_index_delta))
+
+            elif tag == 'replace':
+                i1_index = old_unitids[old_unitid_list[i1 - 1]]['index']
+                i2_index = old_unitids[old_unitid_list[i2 - 1]]['index']
+                update_index_delta = j2 - j1 - i2_index + i1_index
+                inserts.append((i1_index,
+                                new_unitid_list[j1:j2],
+                                i2_index,
+                                update_index_delta))
+        return inserts
+
+    def get_common_dbids(self, opcodes, old_unitid_list, old_unitids):
+        """Returns a set of unit DB ids to be updated.
+
+        :param opcodes: result of new and old unitid lists matching.
+        :param old_unitid_list: old list of UIDs.
+        :param old_unitids: dictionary {'dbid': dbid, 'index': index}
+        :return: a set of unit DB ids to be updated.
+        """
+        update_dbids = set()
+
+        for (tag, i1, i2, j1, j2) in opcodes:
+            if tag == 'equal':
+                update_dbids.update(set(old_unitids[uid]['dbid']
+                                for uid in old_unitid_list[i1:i2]))
```

Comment:
This should be more indented so it is properly aligned with the rest of the set content:

```python
            if tag == 'equal':
                update_dbids.update(set(old_unitids[uid]['dbid']
                                        for uid in old_unitid_list[i1:i2]))
```